test/module-compile-fail.cc

```diff
@@ -0,0 +1,13 @@
+import fmt;
+
+// the detail namespace must be invisible
+using namespace fmt::detail;
+
+#if defined(FMT_HIDE_MODULE_BUGS) && defined(_MSC_FULL_VER) && _MSC_FULL_VER <= 192930035
+// bug in msvc 16.10-pre3:
+// the namespace is visible even when it is neither
+// implicitly nor explicitly exported
+#  error namespace 'fmt::detail' is visible!
+#endif
+
+int main() {}
```

Comment:
I think it would be better to make it a normal test case in `module-test`. You can detect absence of `fmt::detail` just using lookup mechanisms, something like:

```c++
namespace fmt {
  //namespace detail {}
}

namespace detail {
int test_detail_namespace;
}

namespace fmt {
  void test() {
    using namespace detail;
    int success = test_detail_namespace; // fails to compile if fmt::detail is visible
  }
}
```

Reply:
Well, mulling over this suggestion and playing with it a little I have to admit that I don't understand its benefit. May be I'm just dense.

In both implementations, the compiler will stop with a hard error in one of both possible outcomes:

- in my implementation it causes a hard error if everything is fine and the conditional compilation part will substitute it with an `#error` directive if namespace `fmt::detail` erroneously happens to be visible. The test is supposed to fail compilation and `ctest` will flag `module-compile-fail` as "failing" if the compilation succeeds despite the expectation.
The other, 'normal' `module-test` is supposed to compile. The visibility of `fmt::detail` is benign in the sense that it doesn't preclude clients from using the module in full. So it shouldn't affect the other module-related tests.
- in your suggested implementation, the compile failure is not handled by the test infrastructure if namespace `fmt::detail` is visible and no other test in `module-test` is run. I don't think this kind of testing behaviour is helpful.

I've been looking for a SFINAE-friendly option to check for the visibility of a namespace with unknown contents, but there doesn't seem to be one. After all, given a conforming implementation (which msvc 16.10-pre3 isn't and 16.10 likely won't be, too !) this test is meant to defend contributors from unwittingly exposing supposedly inaccessible implementation details. Namespace `fmt::detail` is one of them, afaik there are more. I'm not sure if a check for these additional namespaces like `dragonbox` or `safe_duration_cast` is necessary.

Reply:
The benefit is that the "test" is moved from CMake to normal C++ code. CMake tests are slow and hard to maintain. It's not different from any other compile error so it's fine if other tests are not run.

Reply:
Ok, the support for compile-fail checks is removed, too.